test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs

```diff
@@ -0,0 +1,129 @@
+
+        public RestoreCommandTests(SignCommandTestFixture fixture)
+        {
+            _testFixture = fixture ?? throw new ArgumentNullException(nameof(fixture));
+            _trustedTestCert = SigningTestUtility.GenerateTrustedTestCertificate();
+            _nugetExePath = _testFixture.NuGetExePath;
+        }
+
+        [CIOnlyFact]
+        public async Task Restore_TamperedPackage_FailsAsync()
+        {
+            // Arrange
+            var nupkg = new SimpleTestPackageContext("A", "1.0.0");
+
+            using (var pathContext = new SimpleTestPathContext())
+            using (var testCertificate = new X509Certificate2(_trustedTestCert.Source.Cert))
+            {
+                // Set up solution, project, and packages
+                var solution = new SimpleTestSolutionContext(pathContext.SolutionRoot);
+
+                var projectA = SimpleTestProjectContext.CreateNETCore(
+                    "a",
+                    pathContext.SolutionRoot,
+                    NuGetFramework.Parse("NETStandard2.0"));
+
+                var packageX = new SimpleTestPackageContext("X", "9.0.0");
+                var signedPackagePath = await SignedArchiveTestUtility.AuthorSignPackageAsync(testCertificate, packageX, pathContext.PackageSource);
+                SignedArchiveTestUtility.TamperWithPackage(signedPackagePath);
+
+                projectA.AddPackageToAllFrameworks(packageX);
+                solution.Projects.Add(projectA);
+                solution.Create(pathContext.SolutionRoot);
+
+                var args = new string[]
+                {
+                    projectA.ProjectPath,
+                    "-Source",
+                    pathContext.PackageSource
+                };
+
+                // Act
+                var result = RunRestore(_nugetExePath, pathContext, expectedExitCode: 1, additionalArgs: args);
+                var reader = new LockFileFormat();
+                var lockFile = reader.Read(projectA.AssetsFileOutputPath);
+                var errors = lockFile.LogMessages.Where(m => m.Level == LogLevel.Error);
+                var warnings = lockFile.LogMessages.Where(m => m.Level == LogLevel.Warning);
+            
+                // Assert
+                result.ExitCode.Should().Be(1);
+                result.Errors.Should().Contain(_NU3008);
+                result.AllOutput.Should().Contain($"WARNING: {_NU3027}");
+
+                errors.Count().Should().Be(1);
+                errors.First().Code.Should().Be(NuGetLogCode.NU3008);
+                errors.First().Message.Should().Be(_NU3008Message);
+                errors.First().LibraryId.Should().Be(packageX.ToString());
+
+                warnings.Count().Should().Be(1);
+                warnings.First().Code.Should().Be(NuGetLogCode.NU3027);
+                warnings.First().Message.Should().Be(_NU3027Message);
+                warnings.First().LibraryId.Should().Be("X.9.0.0");
+            }
+        }
```

Comment:
Consider adding more tests to different errors and warnings to be sure they are bubbling up to restore.